### ISSUE TITLE
Hash value for json/net loaded?

### ISSUE THREAD

**Issue by utcke** [Dobiasd/frugally-deep#93]
I think it would be handy for us to have a hash over a loaded model (so I could store, together with the results, some indication of how they were generated - particularly handy for encodings, which tend to be incompatible).
I could simply calculate a hash over the file/string used to initialise the net, but since many files could potentially result in the same net it would be nicer if the net itself could provide such a hash.  Is such a function implemented or, if not, do you see an easy way to get such a hash?

Thanks

Sven

**Comment by Dobiasd** [comment]
Hi Sven,

I guess I don't yet get your request correctly. So I'll try to recap the options I understand:

1. Use unique filenames like `my_cool_model_2018-11-21_16-24`. Use the same base name for the file with the `h5` extension and for the file with the `json` extension.
2. Calculate a hash over the file content and store it, either in the filename or in a separate file next to the model file.
3. Attach some meta information to the internal model name. In Keras you do `model = Model(inputs=x, outputs=y, name='my_model_name')`, so in the `json` file you will find `"architecture"."config"."name": "my_model_name"`. Currently when loading the model it is loaded as the layer name of the `model_layer` that makes up the model. We could add a getter (`const std::string& fdeep::model::name`) to obtain it from C++ application code, that is using the model.
4. Frugally deep itself calculates some kind of hash while converting the `h5` to `json`, and stores it in a new hash field in the json file. We would add a getter for this, like `const std::string& fdeep::model::hash()`.

---

But I'd like to understand the reason behind this. Is the problem you are trying to solve, that you can have many model files and want to remember which file was generated by what python script?

And if fdeep would calculate a hash over to model content and store it with the model, how would that help with your problem?

**Comment by Dobiasd** [comment]
Ah! Maybe you don't necessarily want to store information in the model file. You have some external storage (database, whatever), in which you store some meta information about the model. And there you want to have some kind of unique ID for each model, that you also can find in the model file, to have a safe connection between model file and database entry. Right?

**Comment by utcke** [comment]
Exactly, that is it - I do create encodings (an N-dimensional vector) and store those encodings in an external database.
Now, when the model changes, I want to be able to notice that those embeddings do not match one another, but would need to be recreated.
Being able to access the fdeep::model::hash() would be perfect.
A hash over the file name would not work at all (since I might, in fact usually will, instantiate from a dtring compiled into the executable - but also provide the possibility to instantiate from file instead).
A hash over file content would be possible, but not great, as many different contents would lead to the same memory representation - but I guess this is what fdeep::model::hash() might do?

**Comment by utcke** [comment]
I absolutely agree with you on principle, but as my concern is not two different files creating the same MD5 (the chance of which is actually a lot bigger than people like to believe, but I'm willing to go with 0 for this application - for file-deduplication, on the other hand, your virtually guaranteed a wrong hit for reasonably large filesystems), but two different files creating the same memory representation of the net, which can easily be achived by randomly adding whitespace, for example.  So it would have been nice to somehow calculate a hash for the in-memory representation.

But if there's no easy way to do this (I didn't expect this to be the case, but thought it worth asking), I'll just go the file-contents way...

And please do not take this as a slight in any way - I love fdeep, this is absolutely great and has saved me a ton of work...

**Comment by Dobiasd** [comment]
> the chance of which is actually a lot bigger than people like to believe

I guess you are referring to the hash version of the [https://en.wikipedia.org/wiki/Birthday_problem](Birthday_problem), which also has a surprisingly high probability. I was simply assuming you don't have that many fdeep models stored, that it might become relevant.

---

> but two different files creating the same memory representation of the net, which can easily be achived by randomly adding whitespace, for example.

Do you manually edit the JSON files of the fdeep models, or how can this happen in your case?

---

> So it would have been nice to somehow calculate a hash for the in-memory representation. But if there's no easy way to do this [...]

I guess it would not be that difficult to iterate over the raw data of the layers and build a hash. However for adding this additional complexity (no matter how small) I'd like to first understand (and agree with) the necessity. :)

> And please do not take this as a slight in any way - I love fdeep, this is absolutely great and has saved me a ton of work...
No worry, it's an interesting discussion. And thanks, I'm very happy this hobby project is useful.

**Comment by utcke** [comment]
Damn, need to learn how to cite :-)

Yes, the birthday paradox was what I was referring to.

Yes, I agree the chance of this happening with json models seems negligible, that's why I said I would be willing to pretend the chance on that is 0.

Manually editing the files (in fact just someone accidentally opening the file and the editor doing its automagic) was one of the scenarios I had in mind.  Or perhaps a change in the python script generating those files in a new version of fdeep.

What we will probably end up doing is storing a database of md5sums (or whatever) to version number mappings and auto-generating code from that which will insert the correct model and version number in the cpp file (or, rather, an include file).  Although I would still much rather have a hash over the actual memory content, but I guess it's workable that way too.

But if you should ever implement such a functionality, I would like to know.

**Comment by Dobiasd** [comment]
Mhh, OK, I understand that what you semantically want is a hash over the binary representation of the model. But since it is stored in a non-binary wrapping format (JSON), which can be altered by editors or by transferring it over the wire (thinking about it as the body of a response from a REST service), the JSON hashes might change in cases where this is not desired.

So this hash - hypothetically (:wink:) - would it suffice if there would be `std::string fdeep::model::calculate_hash`, or would you need to have the hash already calculated in Python when convertion from `h5` to `json` with this hash then stored in the resulting `json` file?

Oh, one more question. What is the disadvantage of storing some unique id (timestamp, UUID, whatever) in the model name, i.e., having it as `"architecture"."config"."name": "model_3894563489576348576"` in the `json` file?

**Comment by utcke** [comment]
For my purposes, at least, just having std::string fdeep::model::calculate_hash would work nicely - getting it to match the hash from the json  seems difficult. Then again, if the hash in the json was calculated over content, not the file, that would be fine too - I guess anybody changing the hash in the json gets what he deserves...

Using the file name is a bit difficult as in the way we are using this we actually compile the json into an object containing pointers to the beginning and end of a const char blob with the contents of the file - so the name isn't readily available.

Plus, knowing how things work in practice, I'm virtually certain somebody WILL rename the file, sooner or later... :-)  Experience has show that not only will everything go wrong which could possibly go wrong, but also quite a few of the things you were certain could not possibly go wrong, ever,.,,
And with some clients potentially storing (hundredth of) billions of encodings, you do NOT want a recoding to happen unless REALLY needed :-)

**Comment by Dobiasd** [comment]
The more I think about it, the more sense it seems to make to not have `std::string fdeep::model::calculate_hash() const;` but instead have the hash already stored in the JSON file, meaning calculating it in Python. This might be better because of
- the tooling in Python
- having the hash in the JSON means, it can be read without actually loading the model
- the hash only being calculated once (when the model is converted) instead of each time a C++ client requests it.
- the hash not depending on the C++ client that runs the code. I'm thinking about different floating point precisions etc.

So calculating the hash in `convert_model.py` and adding a field to the JSON object would be the way to go. In C++ of course we could provide a simple `const std::string& fdeep::model::get_hash() const;` then.

---

However I still would like to understand why simply using the model name is not suitable for you. I don't mean the filename. This might indeed change. And if the JSON object is received as the payload from some HTTP GET, there is no filename any more.

But **the model name is not the filename!**

Every fdeep-model-JSON file already contains the following:

```
{
  "architecture": {
    "config": {
      ...
      "name": "my model",
      ...
    }
  }
}
```

`my model` is whatever you name the model in Keras, i.e.,
```python
model = Model(inputs=X, outputs=Y, name='my model')
```

So instead of `my model`, you could use `my model 2018-11-22T19:37:09.873+00:00` or `3ab22a86-ee8e-11e8-8eb2-f2801f1b9fd1` etc.

In C++ we would then provide `const std::string& fdeep::model::get_name() const;`.

What do you think about this?

**Comment by utcke** [comment]
Ok, I missed that point about the model name not being the file name.

Yes, I suppose that _would_ work.  It would mean that every time we re-train the net (more/better data, different data-source etc) we would need to remember to manually change that name though, which once again seems to fall under the heading of "if it's going to fail, it will" - but I suppose just coding the date and time into the model would indeed work nicely, and would be automatic.

Yes, sounds reasonable... (there is a minor problem here in that we might use the same net sampled at different times, but I guess that's sufficiently esoteric not to be worth considering; also, It could be changed manually, but again that would probably classify as ones own fault...).

Thanks pointing that out to me.  How would I access the model-name from inside fdeep?

**Comment by Dobiasd** [comment]
OK, I see, you really want a hash. :D

So I just checked how it could be calculated from a Keras model. It is surprisingly simple:

```python
import hashlib
import numpy as np
from keras.models import Sequential
from keras.layers import Dense, Activation

def calculate_model_hash(model):
    layers = model.layers
    m = hashlib.sha256()
    for layer in layers:
        m.update(np.array(layer.get_weights()).data)
        m.update(layer.name.encode('utf-8'))
    return m.hexdigest()

model = Sequential([
    Dense(5, input_shape=(3,)),
    Activation('relu'),
    Dense(2),
    Activation('softmax'),
])

print(model.name)
model.name = calculate_model_hash(model)
print(model.name)
```

---

> How would I access the model-name from inside fdeep?

With [this fancy new function](https://github.com/Dobiasd/frugally-deep/commit/62138815314d79b56749343ee2cb31c6dcf7ca3f). :wink:

**Comment by utcke** [comment]
Thanks :-)

**Comment by utcke** [comment]
Hmm.

Calling calculate_model_hash several times in a row (without assigning to model.name) results in different hashes - which sort of defeats the idea of a hash...

Not sure why this happens, though...

**Comment by utcke** [comment]
I would guess it hashes the memory address or somesuch - using 

m.update(np.array2string(np.array(a.layers[3].get_weights())).encode('utf-8'))

seems to work consistently

**Comment by Dobiasd** [comment]
:+1:

**Comment by utcke** [comment]
Of course this leaves out much of the middle data, and does not give the same results on Windows and Linux.  Arggghhh.
tostring() and tobyte(), on the other hand, for some reason give slightly different results on each invocation, no idea why.

**Comment by Dobiasd** [comment]
> Of course this leaves out much of the middle data, and does not give the same results on Windows and Linux. Arggghhh.

So you load the model from an h5 file and then build the hash? And this gives different results on Windows and Linux, but same results on subsequent calls on the same OS?

Is this one better?

```python
import base64
import hashlib
import numpy as np
from keras.models import Sequential
from keras.layers import Dense, Activation

def calculate_model_hash(model):
    layers = model.layers
    m = hashlib.sha256()
    for layer in layers:
        m.update(base64.b64encode(np.array2string(np.array(layer.get_weights())).encode('ascii')))
        m.update(layer.name.encode('ascii'))
    return m.hexdigest()

model = Sequential([
    Dense(5, input_shape=(3,)),
    Activation('relu'),
    Dense(2),
    Activation('softmax'),
])

print(model.name)
model.name = calculate_model_hash(model)
print(model.name)
model.name = calculate_model_hash(model)
print(model.name)
```

**Comment by Dobiasd** [comment]
> Of course this leaves out much of the middle data, and does not give the same results on Windows and Linux. Arggghhh.

So you read an `h5` file from disk and build the hash without modifying the model? And this gives different results on Windows and Linux?

How about this one?
```python
import base64
import hashlib
import numpy as np
from keras.models import Sequential
from keras.layers import Dense, Activation

def calculate_model_hash(model):
    layers = model.layers
    m = hashlib.sha256()
    for layer in layers:
        m.update(base64.b64encode(np.array(layer.get_weights())))
        m.update(layer.name.encode('ascii'))
    return m.hexdigest()

model = Sequential([
    Dense(5, input_shape=(3,)),
    Activation('relu'),
    Dense(2),
    Activation('softmax'),
])

print(model.name)
model.name = calculate_model_hash(model)
print(model.name)
model.name = calculate_model_hash(model)
print(model.name)
```

**Comment by utcke** [comment]
It does (different rounding, e.g. I get 0.739451 on Linux, but 0.73945099 on Windows).

The base64 variant for me has the same problem as had tostring() and tobytes():

```
base64.b64encode(np.array(model.layers[3].layers[1].get_weights()))
Out[112]: b'gA8AJAAAAACgAwAkAAAAAA=='
base64.b64encode(np.array(model.layers[3].layers[1].get_weights()))
Out[113]: b'wJEBJAAAAABQ/QEkAAAAAA=='
base64.b64encode(np.array(model.layers[3].layers[1].get_weights()))
Out[114]: b'QJ4BJAAAAADgngEkAAAAAA=='
```

Which is also a surprisingly short string, as 

```
np.array(model.layers[3].layers[1].get_weights())[0].shape
Out[117]: (7, 7, 1, 32)
np.array(model.layers[3].layers[1].get_weights())[1].shape
Out[118]: (32,)
```

So it should be MUCH larger.
I feel like I'm missing something really obvious here, but I'm half tempted to simply run a fixed pattern through the net on initialization and use the output as hash...

**Comment by Dobiasd** [comment]
> but I'm half tempted to [...]

Nooo! :joy:

We'll solve this!

```python
print(np.array(layer.get_weights()).shape)
```
shows the underlying problem.

We have to iterate over the list of weights.

```python
import base64
import hashlib
import numpy as np
from keras.models import Sequential
from keras.layers import Dense, Activation

def calculate_model_hash(model):
    layers = model.layers
    m = hashlib.sha256()
    for layer in layers:
        for weights in layer.get_weights():
            m.update(base64.b64encode(np.array(weights)))
            m.update(layer.name.encode('ascii'))
    return m.hexdigest()

model = Sequential([
    Dense(50, input_shape=(30,)),
    Activation('relu'),
    Dense(2),
    Activation('softmax'),
])

print(model.name)
model.name = calculate_model_hash(model)
print(model.name)
model.name = calculate_model_hash(model)
print(model.name)
```

If one does `print(np.array(weights).shape)` in the inner loop here, the sizes look good this way.

**Comment by utcke** [comment]
I wonder if there's any guarantee on how deeply this can be stacked.

**Comment by Dobiasd** [comment]
> And I suspect that the indentation of "m.update(layer.name.encode('ascii'))" is wrong?

Yes, thanks. I mean, it does no harm in that case, but should be more to the left anyways. :wink:

---

> I wonder if there's any guarantee on how deeply this can be stacked.

For the layers frugally-deep currently supports during export, it seems to never be stacked deeper than that. Otherwise the tests should fail.

However we can make sure this never happens silently by asserting the type:

```python3
import base64
import hashlib
import numpy as np
from keras.models import Sequential
from keras.layers import Dense, Activation

def calculate_model_hash(model):
    layers = model.layers
    m = hashlib.sha256()
    for layer in layers:
        for weights in layer.get_weights():
            assert isinstance(weights, np.ndarray)
            m.update(base64.b64encode(np.array(weights)))
        m.update(layer.name.encode('ascii'))
    return m.hexdigest()

model = Sequential([
    Dense(50, input_shape=(30,)),
    Activation('relu'),
    Dense(2),
    Activation('softmax'),
])

print(model.name)
model.name = calculate_model_hash(model)
print(model.name)
model.name = calculate_model_hash(model)
print(model.name)
```

**Comment by utcke** [comment]
Actually, leaving that line further to the right has the added benefit that empty layers (with no weights) will not contribute to the checksum at all - which might or might not be a good idea, not sure on that yet...

**Comment by Dobiasd** [comment]
If you say "layer names don't mean anything", I think they should be excluded from the hash for all layer, not just weight-less layers?

**Comment by utcke** [comment]
No, My thinking was more along the lines of: when a layer has no weights, it probably doesn't contribute much...

Ok, some further results

```
len(base64.b64encode(weights)) / np.prod(weights.shape)
Out[23]: 5.3341836734693882
```

doesn't have the ring of truth, but 

```
len(weights.tobytes()) / np.prod(weights.shape)
Out[24]: 4.0
```

does.  So I end up with:

```python
def calculate_model_hash(model):
    layers = model.layers
    m = hashlib.sha256()
    for layer in layers:
        for weights in layer.get_weights():
            assert isinstance(weights, np.ndarray)
            m.update(weights.tobytes())
        m.update(layer.name.encode('ascii'))
    return m.hexdigest()
```

And this, in fact, does give the same value for Linux and Windows, so I guess we got that one licked...

**Comment by Dobiasd** [comment]
> ```
> len(base64.b64encode(weights)) / np.prod(weights.shape)
> Out[23]: 5.3341836734693882
> ```
> doesn't have the ring of truth, but
> 
> ```
> len(weights.tobytes()) / np.prod(weights.shape)
> Out[24]: 4.0
> ```
> does.

Yeah, base64 can't provide this property. The `tobytes` version seems to be simpler to me. (And I bet more performant, even without measuring.) 

---

> So I end up with:
> 
> ```python
> def calculate_model_hash(model):
>     layers = model.layers
>     m = hashlib.sha256()
>     for layer in layers:
>         for weights in layer.get_weights():
>             assert isinstance(weights, np.ndarray)
>             m.update(weights.tobytes())
>         m.update(layer.name.encode('ascii'))
>     return m.hexdigest()
> ```
> 
> And this, in fact, does give the same value for Linux and Windows, so I guess we got that one licked...

Awesome! :tada:

**Comment by Dobiasd** [comment]
Do you think we can close this issue?

**Comment by utcke** [comment]
While I still would have liked a hash in C++ I guess we found a suitable solution which at least seems to cover my use-case.
So, yes...

**Comment by Dobiasd** [comment]
What would be the advantage of having it in the C++ part?

I currently only see disadvantages:
1) You can not read the hash from the model json without loading it in fdeep.
2) The hash has to be calculated multiple times for multiple reads.
3) The hash might change depending on the target machine that calculates it (float precision etc.).

**Comment by utcke** [comment]
Well, assuming it hashes over floats in memory, not register, the value should stay the same (just as in Python).
Having to calculate the hash in python means I actually need to remember doing it (i.e. make it part of the script)

However, you are right that it is very convenient to have the hash in the json, and that having to compute it on every model read would be an unnecessary pain.

So I guess we'll just leave it as is. :-)

**Comment by Dobiasd** [comment]
I will not give up before you are completely happy with this. :D

Would calculating the hash automatically in `convert_model.py`

**Comment by utcke** [comment]
Of course it does :-)
